File: Problems & Solutions/2256. Minimum Average Difference/04 December 2022/Approach 1 - Brute Force/Minimum Average Difference.py
class Solution:
    def minimumAverageDifference(self, nums: List[int]) -> int:
        n = len(nums)
        ans = -1
        min_avg_diff = math.inf
        
        for index in range(n):
            left_part_average = 0
            for i in range(index + 1):
                left_part_average += nums[i]
            left_part_average //= (index + 1)
            
            right_part_average = 0
            for j in range(index + 1, n):
                right_part_average += nums[j]
        
            if index != n - 1:
                right_part_average //= (n - index - 1)
            
            curr_difference = abs(left_part_average - right_part_average)
            
            if curr_difference < min_avg_diff:
                min_avg_diff = curr_difference
                ans = index
                
        return ans

File: Problems & Solutions/2256. Minimum Average Difference/04 December 2022/Approach 2 - Prefix Sum/Minimum Average Difference.js
let minimumAverageDifference = function(nums) {
    let n = nums.length;
    let ans = -1;
    let minAvgDiff = Number.MAX_SAFE_INTEGER;

    let prefixSum = Array(n + 1).fill(0);
    let suffixSum = Array(n + 1).fill(0);
    
    for (let index = 0; index < n; ++index) {
        prefixSum[index + 1] = prefixSum[index] + nums[index];
    }

    for (let index = n - 1; index >= 0; --index) {
        suffixSum[index] = suffixSum[index + 1] + nums[index];
    }
    
    for (let index = 0; index < n; ++index) {
        let leftPartAverage = prefixSum[index + 1];
        leftPartAverage = Math.floor(leftPartAverage / (index + 1));

        let rightPartAverage = suffixSum[index + 1];
        if (index != n - 1) {
            rightPartAverage = Math.floor(rightPartAverage / (n - index - 1));
        }

        let currDifference = Math.abs(leftPartAverage - rightPartAverage);
        if (currDifference < minAvgDiff) {
            minAvgDiff = currDifference;
            ans = index;
        }
    }

    return ans;
};